relay-server/src/actors/project_upstream.rs

```diff
@@ -93,32 +103,73 @@ impl ProjectStateChannel {
         self.no_cache = true;
     }
 
-    pub fn send(self, state: ProjectState) {
-        self.sender.send(Arc::new(state)).ok();
+    pub fn attach(&mut self, sender: BroadcastSender<Arc<ProjectState>>) {
+        self.channel.attach(sender)
     }
 
-    pub fn receiver(&self) -> Shared<oneshot::Receiver<Arc<ProjectState>>> {
-        self.receiver.clone()
+    pub fn send(self, state: ProjectState) {
+        self.channel.send(Arc::new(state))
     }
 
     pub fn expired(&self) -> bool {
         Instant::now() > self.deadline
     }
 }
 
+/// This is the [`UpstreamProjectSourceService`] interface.
```

Comment:
_nit_: Since the interface describes the behavior, it would be more straight-forward for a reader to have the bulk of information here. Especially, it would be good to describe what the `UpstreamProjectSource` is and what it can do. I would not link to the service implementation here, but the other way around link to the interface from the service.

Of course, description on internals, such as buffer queues etc can and should remain on the service implementation.